README.md
<!-- markdownlint-disable line-length -->

# fetch-node

Thin wrapper over [`node-fetch`](https://github.com/node-fetch/node-fetch). Sister libray of [`@asd14/fetch-browser`](https://github.com/asd-xiv/fetch-browser).

<!-- vim-markdown-toc GFM -->

* [Install](#install)
* [Initialize](#initialize)
  * [Default headers](#default-headers)
  * [Query string parameters](#query-string-parameters)
* [`GET`](#get)
* [`PATCH`](#patch)
* [`POST`](#post)
* [`DELETE`](#delete)
* [`MULTIPART`](#multipart)
* [Changelog](#changelog)

<!-- vim-markdown-toc -->

## Install

```bash
npm i @asd14/fetch-node
```

## Initialize

```javascript
import { set } from "@asd14/fetch-node"

set({
  // Throws if not set and using relative paths
  baseURL: "http://localhost",
})
```

### Default headers

```javascript
import { set } from "@asd14/fetch-node"

set({
  // Persistent headers
  headers: {
    // Library defaults
    "accept": "application/json",
    "content-type": "application/json",

    // Set JWT for authorized requests
    authorization: "signed-payload-with-base64-over",
  },
})
```

### Query string parameters

There is no built-in way to handle query params but you can set a custom
transform function.

```javascript
import { set } from "@asd14/fetch-node"
import { stringify } from "qs"

set({
  // Throws if query params passed and no stringify function defined
  queryStringifyFn: source =>
    stringify(source, {
      allowDots: true,
      encode: false,
      arrayFormat: "brackets",
      strictNullHandling: true,
    })
})
```

## `GET`

```javascript
import { GET } from "@asd14/fetch-node"

const myIP = await GET("https://api.ipify.org", {
  query: {
    format: "json"
  }
})
// => {"ip":"213.127.80.141"}
```

## `PATCH`

## `POST`

## `DELETE`

## `MULTIPART`

## Changelog

See the [releases section](https://github.com/asd-xiv/fetch-node/releases) for details.


CHANGELOG.md
<!-- markdownlint-disable line-length -->

# Changelog

Releases and changelog are automaticly handled by [semantic-release](https://github.com/semantic-release/semantic-release).

All releases are based on Angular's [Git commit message](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines) patterns.

See the [releases section](https://github.com/asd14-xiv/fetch-node/releases) for details.


tests/index.js
/* eslint-disable new-cap,no-sync */

const { describe } = require("riteway")
const qs = require("qs")

const { set, GET } = require("../src")

set({
  queryStringifyFn: source => qs.stringify(source),
})

describe("fetch-node", async assert => {
  const data = await GET("https://api.ipify.org", {
    query: {
      format: "json",
    },
  })

  assert({
    given: "the burning need to know one's public IP",
    should: "do a GET request to ipify.org",
    actual: {
      hasIpField: typeof data.ip === "string",
    },
    expected: {
      hasIpField: true,
    },
  })
})


src/node.js
/* eslint-disable import/exports-last */

import fetch from "node-fetch"
import {
  pipe,
  reduce,
  startsWith,
  trim,
  when,
  same,
  is,
  isEmpty,
} from "@mutant-ws/m"

import { set } from "./fn.set"
import { filterByValue } from "./fn.filter-by-value"
import { RequestError } from "./fn.request-error"

/**
 * Library options
 */
const props = {
  baseURL: "",
  headers: {},
  queryStringifyFn: null,
}

/**
 * `fetch` wrapper
 *
 * @param  {String} url         API endpoint
 * @param  {String} opt.method  HTTP Method
 * @param  {Object} opt.headers HTTP Headers
 * @param  {Object} opt.body    HTTP Body
 * @param  {Object} opt.query   Query params
 *
 * @return {Promise}            Resolves with response object if code is 20*.
 *                              Reject all other response codes.
 */
const request = (
  url,
  { method, body = {}, headers = {}, query = {}, isFile = false } = {}
) => {
  const reqContent = {
    method,
    headers: filterByValue(is)({
      Accept: "application/json",
      "Content-Type": "application/json",
      ...props.headers,
      ...headers,
    }),
  }

  // Avoid "HEAD or GET Request cannot have a body"
  if (method !== "GET") {
    reqContent.body = isFile ? body : JSON.stringify(body)
  }

  const reqURL = pipe(
    when(
      isEmpty,
      same(url),
      source => `${url}?${props.queryStringifyFn(source)}`
    ),
    trim("/"),
    source => `${props.baseURL}/${source}`
  )(query)

  return fetch(reqURL, reqContent)
    .then(response => {
      const isJSON = startsWith(
        "application/json",
        response.headers.get("Content-Type")
      )

      return Promise.all([response, isJSON ? response.json() : response.text()])
    })
    .then(([response, data]) => {
      /*
       * The Promise returned from fetch() won't reject on HTTP error status
       * even if the response is an HTTP 404 or 500. Instead, it will resolve
       * normally, and it will only reject on network failure or if anything
       * prevented the request from completing.
       */
      if (response.ok) {
        return data
      }

      throw new RequestError(response.statusText, {
        status: response.status,
        body: data,
        url: reqURL,
      })
    })
}

export const useProps = () => [props, set(props)]

export const FILE = (url, { body = {}, headers } = {}) => {
  const form = new FormData()

  return request(url, {
    method: "POST",
    body: pipe(
      Object.entries,
      reduce((acc, [key, value]) => {
        acc.append(key, value)

        return acc
      }, form)
    )(body),
    headers: {
      ...headers,

      // remove content-type header or browser boundery wont get set
      "Content-Type": null,
    },
    isFile: true,
  })
}

export const GET = (url, { query, headers } = {}) =>
  request(url, { method: "GET", query, headers })

export const POST = (url, { body, query, headers } = {}) =>
  request(url, { method: "POST", body, query, headers })

export const PATCH = (url, { body, query, headers } = {}) =>
  request(url, { method: "PATCH", body, query, headers })

export const DELETE = (url, { body, query, headers } = {}) =>
  request(url, { method: "DELETE", body, query, headers })


src/fn.set.js
import { trim, is, isObject } from "@mutant-ws/m"

export const set = props => ({ baseURL, headers, queryStringifyFn }) => {
  if (is(queryStringifyFn)) {
    if (typeof queryStringifyFn === "function") {
      props.queryStringifyFn = queryStringifyFn
    } else {
      throw new TypeError(
        `mutant-fetch: "queryStringifyFn" should be of type function, received ${JSON.stringify(
          queryStringifyFn
        )}`
      )
    }
  }

  if (is(headers)) {
    if (isObject(headers)) {
      props.headers = {
        ...props.headers,
        ...headers,
      }
    } else {
      throw new TypeError(
        `mutant-fetch: "headers" should be of type object, received ${JSON.stringify(
          headers
        )}`
      )
    }
  }

  if (is(baseURL)) {
    if (typeof baseURL === "string") {
      props.baseURL = trim("/")(baseURL)
    } else {
      throw new TypeError(
        `mutant-fetch: "baseURL" should be of type string, received ${JSON.stringify(
          baseURL
        )}`
      )
    }
  }
}


src/browser.js
/* eslint-disable import/exports-last */

import {
  pipe,
  reduce,
  startsWith,
  trim,
  when,
  same,
  is,
  isEmpty,
} from "@mutant-ws/m"

import { set } from "./fn.set"
import { filterByValue } from "./fn.filter-by-value"
import { RequestError } from "./fn.request-error"

/**
 * Library options
 */
const props = {
  baseURL: "",
  headers: {},
  queryStringifyFn: null,
}

/**
 * `fetch` wrapper
 *
 * @param  {String} url         API endpoint
 * @param  {String} opt.method  HTTP Method
 * @param  {Object} opt.headers HTTP Headers
 * @param  {Object} opt.body    HTTP Body
 * @param  {Object} opt.query   Query params
 *
 * @return {Promise}            Resolves with response object if code is 20*.
 *                              Reject all other response codes.
 */
const request = (
  url,
  { method, body = {}, headers = {}, query = {}, isFile = false } = {}
) => {
  const reqContent = {
    method,
    headers: filterByValue(is)({
      Accept: "application/json",
      "Content-Type": "application/json",
      ...props.headers,
      ...headers,
    }),
  }

  // Avoid "HEAD or GET Request cannot have a body"
  if (method !== "GET") {
    reqContent.body = isFile ? body : JSON.stringify(body)
  }

  const reqURL = pipe(
    when(
      isEmpty,
      same(url),
      source => `${url}?${props.queryStringifyFn(source)}`
    ),
    trim("/"),
    source => `${props.baseURL}/${source}`
  )(query)

  return window
    .fetch(reqURL, reqContent)
    .then(response => {
      const isJSON = startsWith(
        "application/json",
        response.headers.get("Content-Type")
      )

      return Promise.all([response, isJSON ? response.json() : response.text()])
    })
    .then(([response, data]) => {
      /*
       * The Promise returned from fetch() won't reject on HTTP error status
       * even if the response is an HTTP 404 or 500. Instead, it will resolve
       * normally, and it will only reject on network failure or if anything
       * prevented the request from completing.
       */
      if (response.ok) {
        return data
      }

      throw new RequestError(response.statusText, {
        status: response.status,
        body: data,
        url: reqURL,
      })
    })
}

export const useProps = () => [props, set(props)]

export const FILE = (url, { body = {}, headers } = {}) => {
  const form = new FormData()

  return request(url, {
    method: "POST",
    body: pipe(
      Object.entries,
      reduce((acc, [key, value]) => {
        acc.append(key, value)

        return acc
      }, form)
    )(body),
    headers: {
      ...headers,

      // remove content-type header or browser boundery wont get set
      "Content-Type": null,
    },
    isFile: true,
  })
}

export const GET = (url, { query, headers } = {}) =>
  request(url, { method: "GET", query, headers })

export const POST = (url, { body, query, headers } = {}) =>
  request(url, { method: "POST", body, query, headers })

export const PATCH = (url, { body, query, headers } = {}) =>
  request(url, { method: "PATCH", body, query, headers })

export const DELETE = (url, { body, query, headers } = {}) =>
  request(url, { method: "DELETE", body, query, headers })


src/fn.set-props.js
const { trim, is, isObject } = require("@asd14/m")

const setProps = props => ({ baseURL, headers, queryStringifyFn }) => {
  if (is(queryStringifyFn)) {
    if (typeof queryStringifyFn === "function") {
      props.queryStringifyFn = queryStringifyFn
    } else {
      throw new TypeError(
        `@asd14/fetch-node: "queryStringifyFn" should be a function, received ${JSON.stringify(
          queryStringifyFn
        )}`
      )
    }
  }

  if (is(headers)) {
    if (isObject(headers)) {
      props.headers = {
        ...props.headers,
        ...headers,
      }
    } else {
      throw new TypeError(
        `@asd14/fetch-node: "headers" should be an object, received ${JSON.stringify(
          headers
        )}`
      )
    }
  }

  if (is(baseURL)) {
    if (typeof baseURL === "string") {
      props.baseURL = trim("/", baseURL)
    } else {
      throw new TypeError(
        `@asd14/fetch-node: "baseURL" should be a string, received ${JSON.stringify(
          baseURL
        )}`
      )
    }
  }
}

module.exports = { setProps }


src/index.js
const fetch = require("node-fetch")
const FormData = require("form-data")
const RFC3986 = require("rfc-3986")
const {
  read,
  pipe,
  reduce,
  startsWith,
  trim,
  when,
  same,
  toLower,
  isEmpty,
} = require("@asd14/m")

const { setProps } = require("./fn.set-props")
const { HTTPError } = require("./fn.http-error")

/**
 * Config
 */
const props = {
  baseURL: "",
  headers: {},
  queryStringifyFn: null,
}

/**
 * `node-fetch` with qs support, default headers and rejects on status
 * outside 200
 *
 * @param  {String} path        API endpoint
 * @param  {String} opt.method  HTTP Method
 * @param  {Object} opt.headers HTTP Headers
 * @param  {Object} opt.body    HTTP Body
 * @param  {Object} opt.query   Query params
 *
 * @return {Promise}            Resolves with response object if code is 20*.
 *                              Reject all other response codes.
 */
const request = (
  path,
  { method, body = {}, headers = {}, query = {} } = {}
) => {
  if (!isEmpty(query) && isEmpty(props.queryStringifyFn)) {
    throw new TypeError(
      `@asd14/fetch-node: ${method}:${path} - Cannot send query params without providing "queryStringifyFn"`
    )
  }

  const isPathURI = new RegExp(RFC3986.uri).test(path)

  if (isEmpty(props.baseURL) && !isPathURI) {
    throw new TypeError(
      `@asd14/fetch-node: ${method}:${path} - Cannot make request with non-absolute path and no "baseURL"`
    )
  }

  // - Remove all undefined values
  // - toLower all keys
  const HEADERS = reduce(
    (acc, [key, value]) =>
      value === undefined
        ? acc
        : {
            ...acc,
            [toLower(key)]: value,
          },
    {},
    Object.entries({
      accept: "application/json",
      "content-type": "application/json",
      ...props.headers,
      ...headers,
    })
  )

  const isReqJSON = pipe(
    read("content-type"),
    startsWith("application/json")
  )(HEADERS)

  const URI = pipe(
    when(
      isEmpty,
      same(path),
      source => `${path}?${props.queryStringifyFn(source)}`
    ),
    trim("/"),
    source => (isPathURI ? source : `${props.baseURL}/${source}`)
  )(query)

  return fetch(URI, {
    method,
    headers: HEADERS,

    // Avoid "HEAD or GET Request cannot have a body"
    ...(method === "GET"
      ? {}
      : { body: isReqJSON ? JSON.stringify(body) : body }),
  })
    .then(response => {
      const isResJSON = startsWith("application/json")(
        response.headers.get("Content-Type")
      )

      return Promise.all([
        response,
        isResJSON ? response.json() : response.text(),
      ])
    })
    .then(([response, data]) => {
      /*
       * The Promise returned from fetch() won't reject on HTTP error status
       * even if the response is an HTTP 404 or 500. Instead, it will resolve
       * normally, and it will only reject on network failure or if anything
       * prevented the request from completing.
       */
      if (response.ok) {
        return data
      }

      throw new HTTPError(response.statusText, {
        status: response.status,
        body: data,
        path: URI,
      })
    })
}

const set = setProps(props)

const GET = (url, { query, headers } = {}) =>
  request(url, { method: "GET", query, headers })

const POST = (url, { body, query, headers } = {}) =>
  request(url, { method: "POST", body, query, headers })

const PATCH = (url, { body, query, headers } = {}) =>
  request(url, { method: "PATCH", body, query, headers })

const DELETE = (url, { body, query, headers } = {}) =>
  request(url, { method: "DELETE", body, query, headers })

const MULTIPART = (url, { body = {}, headers } = {}) => {
  const form = new FormData()

  return request(url, {
    method: "POST",
    body: reduce((acc, [key, value]) => {
      acc.append(key, value)

      return acc
    }, form)(Object.entries(body)),
    headers: {
      ...form.getHeaders(),
      ...headers,
    },
  })
}

module.exports = {
  set,
  GET,
  POST,
  PATCH,
  DELETE,
  MULTIPART,
}
